File: LeetCodePractice/src/bst/KthSmallestElementInaBST.java
package bst;

import java.util.Stack;

/*Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

Example
Given root = {1,#,2}, k = 2, return 2.

Challenge
What if the bst is modified (insert/delete operations) often and you need to find the kth smallest frequently?
How would you optimize the kthSmallest routine?

Notice
You may assume k is always valid, 1 ≤ k ≤ bst's total elements.

analysis:
leverage inorder iterative traversal
*/
public class KthSmallestElementInaBST {
	/**
     * @param root: the given bst
     * @param k: the given k
     * @return: the kth smallest element in bst
     */
    public int kthSmallest(TreeNode root, int k) {
    	//Inorder traversal
    	if(root == null){
    		return -1;
    	}
    	int count = 0;
    	TreeNode cur = root;
    	Stack<TreeNode> stack = new Stack<>();
    	while(cur != null || !stack.isEmpty()){
    		if(cur != null){
    			stack.push(cur);
    			cur = cur.left;
    		} else {
    			cur = stack.pop();
    			count++;
        		if(count == k){
        			return cur.val;
        		}
        		cur = cur.right;
    		}
    	}
    	return -1;
    }
}

File: LeetCodePython/KthSmallestElementInaBST.py
"""
Given the root of a binary search tree, and an integer k, return the kth smallest value (1-indexed) of all the values of the nodes in the tree.

Example 1:
Input: root = [3,1,4,null,2], k = 1
Output: 1

Example 2:
Input: root = [5,3,6,2,4,null,null,1], k = 3
Output: 3


Constraints:
The number of nodes in the tree is n.
1 <= k <= n <= 10^4
0 <= Node.val <= 10^4


Follow up: If the BST is modified often (i.e., we can do insert and delete operations) and you need to find the kth smallest frequently, how would you optimize?

hint:
1 Try to utilize the property of a BST.
2 Try in-order traversal.
3 What if you could modify the BST node's structure?
4 The optimal runtime complexity is O(height of BST).
"""
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class KthSmallestElementInaBST:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                k -= 1
                if k == 0:
                    return root.val
                root = root.right
        return -1
